File: src/modules/graphs/1_Representation/AdjacencyList.ts
/**
0 (Alice) -> 1 (Bob)
1 (Bob) -> 0 (Alice) | 2 (Charlie)
2 (Charlie) -> 1 (Bob) | 3 (David)
3 (David) -> 2 (Charlie) | 4 (Emma)
4 (Emma) -> 3 (David)
 */

class GraphAdjacencyList {
    public vertices: number;
    public list: number[][];
    constructor(vertices: number) {
        this.vertices = vertices;
        this.list = new Array(vertices).fill(null).map(() => new Array());
    }
    addEdge(source: number, destination: number) {
        this.list[source].push(destination);
        this.list[destination].push(source);
    }
    displayGraph(){
        for(let i=0; i<this.vertices; i++){
            console.log(i, '->', this.list[i].join(' '));
        }
    }
}

const vertices = 5;
const graph = new GraphAdjacencyList(vertices);

graph.addEdge(0, 1);
graph.addEdge(0, 2);
graph.addEdge(1, 3);
graph.addEdge(2, 4);

graph.displayGraph();

export { }

File: src/modules/graphs/1_Representation/AdjacencyList.go
package main

import "fmt"

// GraphAdjacencyList represents a graph using an adjacency list
type GraphAdjacencyList struct {
	vertices int
	list     [][]int
}

// NewGraph creates a new graph with the specified number of vertices
func NewGraph(vertices int) *GraphAdjacencyList {
	g := &GraphAdjacencyList{
		vertices: vertices,
		list:     make([][]int, vertices),
	}
	for i := range g.list {
		g.list[i] = []int{}
	}
	return g
}

// AddEdge adds an edge between the source and destination vertices
func (g *GraphAdjacencyList) AddEdge(source, destination int) {
	g.list[source] = append(g.list[source], destination)
	g.list[destination] = append(g.list[destination], source)
}

// DisplayGraph displays the adjacency list representation of the graph
func (g *GraphAdjacencyList) DisplayGraph() {
	for i := 0; i < g.vertices; i++ {
		fmt.Printf("%d -> %v\n", i, g.list[i])
	}
}

func main() {
	vertices := 5
	graph := NewGraph(vertices)

	graph.AddEdge(0, 1)
	graph.AddEdge(0, 2)
	graph.AddEdge(1, 3)
	graph.AddEdge(2, 4)

	graph.DisplayGraph()
}
